src/workprover/__init__.py

```diff
@@ -0,0 +1,238 @@
+import sys
+import multiprocessing
+import time
+import struct
+import threading
+import Queue
+import collections
+
+import utils
+import dumbsolver
+import forkingsolver
+import fastsolver
+import gpusolver
+
+timeout = .5
+
+class Stop(Exception):
+    pass
+
+class Task(object):
+    previous = None
+    next = None
+
+    def __init__(self, headlessPayload, TTL, expiryTime, target):
+        self.headlessPayload = headlessPayload
+        self.TTL = TTL
+        self.expiryTime = expiryTime
+        self.target = target
+
+class WorkProver(threading.Thread):
+    def __init__(self, codePath, GPUVendors, seed, statusUpdated):
+        super(self.__class__, self).__init__()
+
+        self.availableSolvers = {
+            "dumb": dumbsolver.DumbSolver(codePath)
+        }
+
+        # Comment from the previous version:
+
+        # on my (Peter Surda) Windows 10, Windows Defender
+        # does not like this and fights with PyBitmessage
+        # over CPU, resulting in very slow PoW
+        # added on 2015-11-29: multiprocesing.freeze_support() doesn't help
+
+        if not hasattr(sys, "frozen") or sys.frozen == "macosx_app":
+            self.availableSolvers["forking"] = forkingsolver.ForkingSolver(codePath)
+
+        try:
+            self.availableSolvers["fast"] = fastsolver.FastSolver(codePath)
+        except:
+            pass
+
+        try:
+            self.availableSolvers["gpu"] = gpusolver.GPUSolver(codePath, GPUVendors)
+        except:
+            pass
+
+        try:
+            self.defaultParallelism = multiprocessing.cpu_count()
+        except NotImplementedError:
+            self.defaultParallelism = 1
+
+        self.seed = seed
+        self.roundsCounter = 0
+        self.statusUpdated = statusUpdated
+
+        self.commandsQueue = Queue.Queue()
+        self.resultsQueue = Queue.Queue()
+
+        self.solverName = None
+        self.solver = None
+
+        self.lastTime = utils.getTimePoint()
+        self.timedIntervals = collections.deque()
+        self.speed = 0
+
+        self.tasks = {}
+        self.currentTaskID = None
+
+    def notifyStatus(self):
+        if self.statusUpdated is None:
```

Comment:
It's less acurate.

`self.statusUpdated` must be either a function or a `None`. If it's 0 or an empty string, it would be a programming error and further call would rise an exception making the error noticable.